stl/inc/xutility

```diff
@@ -281,23 +271,37 @@ void _Default_construct_in_place(_Ty& _Obj) noexcept(is_nothrow_default_construc
     ::new (_Voidify_iter(_STD addressof(_Obj))) _Ty;
 }
 
-_EXPORT_STD template <class _Ty>
-struct pointer_traits {
+template <class _Ty, class _Elem>
+struct _Ptr_traits_base {
     using pointer         = _Ty;
-    using element_type    = typename _Get_element_type<_Ty>::type;
+    using element_type    = _Elem;
     using difference_type = typename _Get_ptr_difference_type<_Ty>::type;
 
     template <class _Other>
     using rebind = typename _Get_rebind_alias<_Ty, _Other>::type;
 
-    using _Reftype = conditional_t<is_void_v<element_type>, char, element_type>&;
+    using _Reftype = conditional_t<is_void_v<_Elem>, char, _Elem>&;
 
     _NODISCARD static _CONSTEXPR20 pointer pointer_to(_Reftype _Val) noexcept(
         noexcept(_Ty::pointer_to(_Val))) /* strengthened */ { // Per LWG-3454
         return _Ty::pointer_to(_Val);
     }
 };
 
+template <class, class = void, class = void>
+struct _Ptr_traits_sfinae_layer {};
+
+template <class _Ty, class _Uty>
+struct _Ptr_traits_sfinae_layer<_Ty, _Uty, void_t<typename _Get_first_parameter<_Ty>::type>>
+    : _Ptr_traits_base<_Ty, typename _Get_first_parameter<_Ty>::type> {};
+
+template <class _Ty>
+struct _Ptr_traits_sfinae_layer<_Ty, void_t<typename _Ty::element_type>, void>
+    : _Ptr_traits_base<_Ty, typename _Ty::element_type> {};

```

Comment:
So this seems to basically do as follows:

The third template specialization (right line 298) is the most specific of the three, so if it's valid, it'll be chosen, and our `element_type` will be `_Ty::element_type`.

If the third template specialization is not chosen, because `_Ty::element_type` is not a type name, then the second (right line 294) is still more specific than the first (right line 291), and if it is valid, we'll choose it.

Finally, otherwise, if `_Ty` is not a class template with only type arguments, we fall back to having no members.

This makes sense, and I think correctly implements the wording in the standard.

Reply:
I don't necessarily believe this is true, since if `typename _Get_first_parameter<_Ty>::type` and `typename _Ty::element_type` are both well formed (which they often are, like `allocator<_Ty>`), then both specializations are valid and equally specialized.